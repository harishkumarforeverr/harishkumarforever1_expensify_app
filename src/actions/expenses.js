import { v4 as uuid } from 'uuid';
import database from "../firebase/firebase";
// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
});
//StartAddExpense
export const StartAddExpense=(expense={})=>{
    return (dispatch)=>{
          const { description = "",  note = "", amount = 0, createdAt = 0  }=expense
          const addingExpensesObject={description,note,amount,createdAt };
          return database.ref("expenses").push(addingExpensesObject).then((snapshot)=>{
              dispatch(addExpense({
                  id:snapshot.key,
                  ...addingExpensesObject
              }))
          });
    }
}

// REMOVE_EXPENSE

export const removeExpense = ({ id } = {}) => ({
    type: "REMOVE_EXPENSE",
    id
});
// EDIT_EXPENSE

export const editExpanse = (id, update) => ({
    type: "EDIT_EXPANSE",
    id,
    update
})

//SET expenses
export const setExpenses=(expense)=>({
    type:"SET_EXPENSES",
    expense 
})

export const startSetExpenses=()=>{
    return (dispatch)=>{
        return database.ref("expenses").once("value").then((snapshot)=>{   
            const expenses=[];
            snapshot.forEach((childSnapShot)=>{     
                expenses.push({
                    id:childSnapShot.key, 
                    ...childSnapShot.val()
              })
            })
            dispatch(setExpenses(expenses));
        })
    }
}