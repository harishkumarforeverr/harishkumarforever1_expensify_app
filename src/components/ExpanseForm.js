import moment from 'moment';
import React from 'react'; 
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';

export default class ExpanseForm extends React.Component{
    constructor(props){   
       super(props);
       this.state={ 
            description:props.expense ? props.expense.description:"",
            note:props.expense ? props.expense.note:"",
            amount:props.expense ? (props.expense.amount/100).toString() :"",
            createdAt:props.expense ? moment(props.expense.createdAt) : moment(),
            focused:false,
            formErrorMessage:""
        };
    }
    onDescriptionChange=(e)=>{ 
        this.setState(()=>{
            const description=e.target.value;
            return {
                description
            }
        })
    } 
    onNotesChange=(e)=>{
        this.setState(()=>{
            const note=e.target.value
            return {
                note
            }
        })
    }
    onAmountChange=(e)=>{ 
        const amount=e.target.value ;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>{
                return {
                    amount
                }
            });
        }  
    } 
    onDateChange = (createdAt)=>{
        if(createdAt){
            this.setState({createdAt})
        }
    }
    onSubmit=(e)=>{
      e.preventDefault(); 
      if(this.state.description && this.state.amount){ 
        this.setState(()=>{
            return {
                formErrorMessage:""
            }
        }); 
       this.props.onSubmit({
           description:this.state.description,
           note:this.state.note,
           amount:parseFloat(this.state.amount,10)*100,
           createdAt:this.state.createdAt.valueOf() 
       });       
      }
      else{ 
        this.setState(()=>{
            return {
                formErrorMessage:"please provide Description and Amount (mandatory.)"
            }
        }) 
        
      }
    }
    render(){
        return (   
            <form className="form" onSubmit={this.onSubmit}>
             {this.state.formErrorMessage && <h4 className="form__error"> {this.state.formErrorMessage} </h4>}
              <input 
              className="text-input"
               type="text"
               value={this.state.description}
               placeholder="add your expanses" 
               autoFocus
               onChange={this.onDescriptionChange}
               />
               <input 
               className="text-input"
               type="number" 
               placeholder="add amount" 
               value={this.state.amount}
               onChange={this.onAmountChange}
               />
               <SingleDatePicker 
               date={this.state.createdAt}
               onDateChange={this.onDateChange}
               focused={this.state.focused}
               onFocusChange={ ({focused}) =>this.setState({ focused }) }
               numberOfMonths={1}
               isOutsideRange={()=>false}
               />
               <textarea
               className="textarea"
               placeholder="add a notes to your expenses (Optional !)"
               value={this.state.note}
               onChange={this.onNotesChange}
               >

               </textarea>
               <div> 
                    <button class="button"> Save Expense </button>
               </div>
            </form> 
        );
    }
}