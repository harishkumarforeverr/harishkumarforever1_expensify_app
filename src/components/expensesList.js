import React from 'react';
import { connect } from 'react-redux';
import ExpensesListItems from "./expensesListItems"
import selectedExpenses from "../selectors/expenses";
export const ExpensesList=(props)=>(
  <div>          
        {
            props.expenses.length===0?(
                <h4>Zero exoenses  </h4>
            ):( 
               props.expenses.map((expense)=>{
                  return (<ExpensesListItems key={expense.id} {...expense}  />);
               })
            )
        }

  </div>
);

const mapStateToProps= (state)=>{
    return ({
    expenses:selectedExpenses(state.expenses,state.filters)        
    })
}
 export default connect(mapStateToProps)(ExpensesList);
