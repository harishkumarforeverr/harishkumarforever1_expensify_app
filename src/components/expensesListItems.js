import React from 'react';
import {Link} from "react-router-dom"

const ExpensesListItems=({id,description ,amount,createdAt,note})=>{   
    return (
        <div> 
            <Link to={`/edit/${ id }`}> <h3 className="expenses">   {description} </h3> </Link>
             <p>  {amount} -- {createdAt} </p>
             <hr/>
        </div>
    );
}


export default ExpensesListItems;