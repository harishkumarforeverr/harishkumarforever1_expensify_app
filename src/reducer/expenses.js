
// expanse Reducer 
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {  
        case "ADD_EXPENSE":
            return ([
                ...state,
                action.expense
            ]);
        case "REMOVE_EXPENSE":
            return state.filter((expense) => expense.id !== action.id);
        case "EDIT_EXPANSE" :
            return state.map((expanse)=>{
               if(expanse.id===action.id){
                   return {
                    ...expanse,
                    ...action.update
                   }
               }
               else{
                   return expanse;
               }
            });
        case "SET_EXPENSES":
            return action.expense; 
        default: return state
    }
}
