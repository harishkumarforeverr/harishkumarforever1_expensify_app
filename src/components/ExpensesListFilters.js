import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter} from "../actions/filters";
import {sortByDate, sortByAmount,setStartDate,setEndDate} from "../actions/filters";
import {DateRangePicker} from "react-dates";
import moment from "moment";
// from i took the file from css  and placed it in src/app.js 

export class ExpensesListFilters extends React.Component{ 
      state={
            focusedInput:null
        } 
    onDatesChange=({startDate,endDate})=>{
        this.props.setStartDate(startDate) ;
        this.props.setEndDate(endDate) ;            
    }
    onTextChange=(e)=>{
       this.props.setTextFilter(e.target.value);   
    }
    onSortByChange=(e)=>{ 
        if(e.target.value=="amount") {
            this.props.sortByAmount();
        }   
        else if(e.target.value=="date"){
            this.props.sortByDate();
        }         
   }
    render(){ 
        return (
       <div>
            <div className="input-group"> 
                <div className="input-group__item"> 
                     <input 
                       className="text-input" 
                        type="text" 
                        value={this.props.filters.text}
                        onChange={this.onTextChange}
                        autoFocus
                        placeholder="search expenses(items)"
                        />
                </div>
                <div className="input-group__item"> 
                       <select 
                       className="select"
                        value={this.props.filters.sortBy}
                        onChange={this.onSortByChange}  
                        >
                        <option value="amount">Amount</option>
                        <option value="date">Date</option>
                        </select>
                </div>
                <div className="input-group__item"> 
                        <DateRangePicker 
                        startDate={this.props.filters.startDate}
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.focusedInput}
                        onFocusChange={focusedInput=>this.setState({focusedInput})}
                        startDateId="DateRangePicker_startDateId"
                        endDateId="DateRangePicker_endDateId"
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                        showClearDates={true}
                        />      
                </div> 
               
              
          </div>
       </div>
    );
    }
}
 

const mapStateToProps=(state)=>{
    return {
        filters:state.filters
    }
}
const mapDispatchToProps=(dispatch)=>{ 
    return { 
        setStartDate  : (startDate)=>{ dispatch(setStartDate(startDate))  },
        setEndDate: (endDate)=>{ dispatch(setEndDate(endDate))},
        setTextFilter: (value)=>{dispatch(setTextFilter(value))},
        sortByAmount: ()=>{dispatch(sortByAmount())},
        sortByDate : ()=>{dispatch(sortByDate())},
    } 
}
export default connect(mapStateToProps,mapDispatchToProps)(ExpensesListFilters);