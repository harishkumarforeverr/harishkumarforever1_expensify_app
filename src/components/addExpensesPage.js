import React from "react";  
import ExpanseForm from "./ExpanseForm";
import { addExpense } from "../actions/expenses";
import {connect} from 'react-redux';

export class AddExpensesPage extends React.Component{
    addExpense=(expense)=>{ 
        this.props.addExpense(expense)
        this.props.history.push("/");
    }
    render(){
        return (
            <div>
                <h1>Add Expanse </h1>
                <ExpanseForm 
                onSubmit={this.addExpense}
                 />
            </div>
        );
    }
}
const mapDispatchToState=(dispatch)=>{
    return {
        addExpense:(expense)=>{  dispatch(addExpense(expense)); }
    }
}
 
export default connect(undefined,mapDispatchToState)(AddExpensesPage);