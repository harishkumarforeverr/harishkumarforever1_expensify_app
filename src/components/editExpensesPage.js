import React from 'react';
import {connect} from 'react-redux';
import ExpanseForm from "./ExpanseForm";
import {startEditExpense} from "../actions/expenses";
import {startRemoveExpense} from '../actions/expenses';


export class EditExpensesPage extends React.Component{ 
    startEditExpense=(expense)=>{  
        this.props.startEditExpense(this.props.expense.id, expense );  
        this.props.history.push("/");
    }
    onClick=()=>{  
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push("/");
       }
    render(){
        return (
            <div>
                <ExpanseForm 
                    expense={this.props.expense}
                    onSubmit={this.startEditExpense}
                 /> 
                 <button
                  onClick= {this.onClick } 
                  >Remove
                  </button>
            </div>
        );
    }
}
const mapDispatchToState=(dispatch)=>{
    return { 
    startEditExpense : (id,expense)=>{   dispatch(startEditExpense(id,expense))   },
    startRemoveExpense:(id )=>{ dispatch(startRemoveExpense({ id }))  }
    }
}

const mapStateToProps=(state, props)=>{
    return {
        expense : state.expenses.find((expense)=> expense.id===props.match.params.id )
    }
}
export default connect(mapStateToProps,mapDispatchToState)(EditExpensesPage);