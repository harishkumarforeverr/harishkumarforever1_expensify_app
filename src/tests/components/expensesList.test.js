import React from 'react';
import {shallow} from "enzyme";
import { ExpensesList } from "../../components/expensesList";
import expenses from "../fixtures/expenses";

test("test/components/expensesList should render expensesList correctly ",()=>{
     const wrapper=shallow(<ExpensesList expenses={expenses} />);
     expect(wrapper).toMatchSnapshot();
});


test("test/components/expensesList should render expensesList correctly ",()=>{
    const wrapper=shallow(<ExpensesList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});