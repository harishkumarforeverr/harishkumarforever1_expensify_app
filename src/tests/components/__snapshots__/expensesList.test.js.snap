// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/components/expensesList should render expensesList correctly  1`] = `
<div>
  <ExpensesListItems
    amount={30}
    createdAt={0}
    description="buy long note books"
    id={0}
    key="0"
    notes="buying the gun"
  />
  <ExpensesListItems
    amount={2500}
    createdAt={-345600000}
    description="my room rent"
    id={1}
    key="1"
    notes="my house rent"
  />
  <ExpensesListItems
    amount={600}
    createdAt={345600000}
    description="phone Reacharge"
    id={2}
    key="2"
    notes="took creatied Card for my mom from SBI bank"
  />
</div>
`;

exports[`test/components/expensesList should render expensesList correctly  2`] = `
<div>
  <h4>
    Zero exoenses  
  </h4>
</div>
`;
